thirtyinch/src/main/java/net/grandcentrix/thirtyinch/internal/PresenterSavior.java

```diff
@@ -32,36 +39,261 @@
+                    // detected Activity finish, no new Activity instance will be created
+                    // with savedInstanceState, clear saved presenters
+                    cleanupAfterFinish(activity);
+                } else {
+                    // don't leak old activity instances
+                    // scopeId is saved in savedInstanceState of finishing Activity.
+                    mScopeIdForActivity.remove(activity);
+                }
+            }
+
+            @Override
+            public void onActivityPaused(final Activity activity) {
+
+            }
+
+            @Override
+            public void onActivityResumed(final Activity activity) {
+
+            }
+
+            @Override
+            public void onActivitySaveInstanceState(final Activity activity,
+                    final Bundle outState) {
+                saveScopeId(activity, outState);
+            }
+
+            @Override
+            public void onActivityStarted(final Activity activity) {
+
+            }
+
+            @Override
+            public void onActivityStopped(final Activity activity) {
+
+            }
+        };
+    }
+
+    public void saveScopeId(final Activity activity, final Bundle outState) {
+        String id = mScopeIdForActivity.get(activity);
+        if (id == null) {
+            id = UUID.randomUUID().toString();
+            mScopeIdForActivity.put(activity, id);
+        }
+        outState.putString(TI_ACTIVITY_PRESENTER_SCOPE_KEY, id);
+    }
+
+    private synchronized ActivityScopedPresenters getScope(final Activity activity) {
+        final String id = mScopeIdForActivity.get(activity);
+        if (id != null) {
+            return mScopes.get(id);
+        }
+
+        return null;
+    }
+
+    private synchronized ActivityScopedPresenters getScopeOrCreate(final Activity activity) {
+        if (mLifecycleCallbacks == null) {
+            mLifecycleCallbacks = getActivityLifecycleCallbacks();
+            TiLog.v(TAG, "registering lifecycle callback");
+            final Application application = activity.getApplication();
+            application.registerActivityLifecycleCallbacks(mLifecycleCallbacks);
+        }
+
+        final String scopeId = mScopeIdForActivity.get(activity);
+        final ActivityScopedPresenters scope;
+
+        if (scopeId == null) {
+            final String id = UUID.randomUUID().toString();
+            mScopeIdForActivity.put(activity, id);
+            scope = new ActivityScopedPresenters();
+            mScopes.put(id, scope);
+        } else {
+            final ActivityScopedPresenters gotScope = mScopes.get(scopeId);
+            if (gotScope == null) {
+                scope = new ActivityScopedPresenters();
+                mScopes.put(scopeId, scope);
+            } else {
+                scope = gotScope;
+            }
+        }
+
+        return scope;
+    }
+
+    private void printRemainingStore() {
```

Comment:
This is only for debugging right? 🤔 
I would prefer to add these info to the method name.
I also recomment to create a static boolean which can be toggled manually to identify if we have to run this code or not...

```
privat static final boolean DEBUG = true;

private void printDebugRemainingStore() {
    if(DEBUG) {
        // Run that code
    }
}
```